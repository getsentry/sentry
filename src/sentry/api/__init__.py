from .analytics import *  # NOQA
from .client import ApiClient

client = ApiClient()
