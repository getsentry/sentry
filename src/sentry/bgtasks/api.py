from __future__ import absolute_import

import six
import time
import random
import logging
import threading
from contextlib import contextmanager

from django.conf import settings


logger = logging.getLogger("sentry.bgtasks")
tasks = {}


def bgtask(roles=None, interval=60):
    def decorator(f):
        return BgTask(callback=f, roles=roles, interval=interval)

    return decorator


class BgTask(object):
    def __init__(self, callback, roles=None, interval=60):
        self.callback = callback
        self.roles = roles or []
        self.interval = interval
        self.running = False

    @property
    def name(self):
        return "%s:%s" % (self.callback.__module__, self.callback.__name__)

    def run(self):
        if self.running:
            return
        self.running = True

        next_run = time.time() + self.interval * random.random()
        while self.running:
            now = time.time()
            if now >= next_run:
                try:
                    self.callback()
                except Exception:
                    logging.error("bgtask.failed", exc_info=True, extra=dict(task_name=self.name))
                next_run = now + self.interval
            time.sleep(1.0)

    def reconfigure(self, cfg):
        if "roles" in cfg:
            self.roles = cfg["roles"]
        if "interval" in cfg:
            self.interval = cfg["interval"]

    def spawn_daemon(self):
        if self.running:
            return
        logger.info("bgtask.spawn", extra=dict(task_name=self.name))
        t = threading.Thread(target=self.run)
        t.setDaemon(True)
        t.start()

    def stop(self):
        logger.info("bgtask.stop", extra=dict(task_name=self.name))
        self.running = False


def get_task(task_name):
    module, task_cls = task_name.split(":", 1)
    mod = __import__(module, None, None, [task_cls])
    return getattr(mod, task_cls)


def spawn_bgtasks(role):
    for import_name, cfg in six.iteritems(settings.BGTASKS):
        task = get_task(import_name)
        # This is already running
        if task.name in tasks:
            continue
        task.reconfigure(cfg)
        if role not in task.roles:
            continue
        task.spawn_daemon()
        tasks[task.name] = task


def shutdown_bgtasks():
    for task_name, task in list(six.iteritems(tasks)):
        task.stop()
        tasks.pop(task_name, None)


@contextmanager
def managed_bgtasks(role):
    spawn_bgtasks(role)
    try:
        yield
    finally:
        shutdown_bgtasks()
