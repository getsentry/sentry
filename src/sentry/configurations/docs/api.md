# Configurations API

Host: https://sentry.io/api/0

**Authors.**

@cmanallen

## Configuration [/projects/<organization_id_or_slug>/<project_id_or_slug>/configuration/]

### Get Configuration [GET]

Retrieve the project's configuration.

**Attributes**

| Column             | Type   | Description                     |
| ------------------ | ------ | ------------------------------- |
| id                 | number | Project id.                     |
| sample_rate        | number | A number value between 0 and 1. |
| traces_sample_rate | number | A number value between 0 and 1. |
| user_config        | any    | Arbitrary user supplied JSON.   |

- Response 200

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 1.0,
      "traces_sample_rate": 0.5,
      "user_config": {
        "hello": "world"
      }
    }
  }
  ```

### Update Configuration [PATCH]

Update the project's configuration.

- Request

  ```json
  {
    "data": {
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

- Response 201

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

### Delete Configuration [DELETE]

Delete the project's configuration.

- Response 204
