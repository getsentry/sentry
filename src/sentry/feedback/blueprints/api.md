# Feedback API

Host: https://sentry.io/api/0

**Authors.**

@cmanallen

**How to read this document.**

This document is structured by resource with each resource having actions that
can be performed against it. Every action that either accepts a request or
returns a response _will_ document the full interchange format. Clients may opt
to restrict response data or provide a subset of the request data. The API may
or may not accept partial payloads.

## Feedback Index [/organizations/<organization_slug>/feedback/]

- Parameters

  - field (optional, string)
  - environment (optional, string)
  - project (optional, string)
  - sort (optional, string) Default: -timestamp Members
    - projectId
    - -projectId
    - timestamp
    - -timestamp
  - statsPeriod (optional, string) - A positive integer suffixed with a unit
    type. Default: 7d Members
    - s
    - m
    - h
    - d
    - w
  - start (optional, string) - ISO 8601 format (`YYYY-MM-DDTHH:mm:ss.sssZ`)
  - end (optional, string) - ISO 8601 format. Required if `start` is set.
  - utc (optiona, 'true') - Whether start/end should use the UTC timezone
  - cursor (optional, str)
  - per_page (optional, number) Default: 10
  - offset (optional, number) Default: 0
  - query (optional, string) - Search query with space-separated field/value
    pairs. ie: `?query=browser:Firefox`.
  - queryReferrer(optional, string) - Specify the page which this query is being
    made from. Used for cross project query on issue replays page. Pass
    `queryReferrer=replayUserFeedback` for this query.

### Browse Feedback [GET]

Retrieve a collection of feedback items.

**Attributes**

| Column            | Type             | Description                                         |
| ----------------- | ---------------- | --------------------------------------------------- |
| browser           | object           | -                                                   |
| browser.name      | optional[string] | -                                                   |
| browser.version   | optional[string] | -                                                   |
| contact_email     | string           | The contact email of the user writing feedback.     |
| device            | object           | -                                                   |
| device.brand      | optional[string] | -                                                   |
| device.family     | optional[string] | -                                                   |
| device.model      | optional[string] | -                                                   |
| device.name       | optional[string] | -                                                   |
| dist              | optional[string] |                                                     |
| environment       | optional[string] |                                                     |
| feedback_id       | string           |                                                     |
| locale            | object           | Browser locale information.                         |
| locale.lang       | string           | The language preference of the user's browser.      |
| locale.timezone   | string           | The timezone the feedback was submitted from.       |
| message           | string           | The message written by the user providing feedback. |
| os                | object           | -                                                   |
| os.name           | optional[string] | -                                                   |
| os.version        | optional[string] | -                                                   |
| platform          | string           |                                                     |
| project_id        | string           |                                                     |
| release           | optional[string] |                                                     |
| replay_id         | optional[string] | The id of a running replay.                         |
| sdk               | object           | SDK information.                                    |
| sdk.name          | string           | -                                                   |
| sdk.version       | string           | -                                                   |
| status            | string           | One of: resolved, unresolved.                       |
| tags              | object           | Mapping of key, value pairs.                        |
| timestamp         | string           | ISO-8061 formatted UTC datetime.                    |
| url               | string           | The page the feedback was triggered on.             |
| user              | object           | The authorized user's information.                  |
| user.display_name | optional[string] | -                                                   |
| user.email        | optional[string] | -                                                   |
| user.id           | optional[string] | -                                                   |
| user.ip           | optional[string] | Same search field as Events                         |
| user.username     | optional[string] | -                                                   |

- Response 200

  - Headers

    ```
    X-Hits: 42
    ```

  - Body
    ```json
    {
      "data": [
        {
          "browser": {
            "name": "Chome",
            "version": "103.0.38"
          },
          "contact_email": "colton.allen@sentry.io",
          "device": {
            "brand": "Apple",
            "family": "iPhone",
            "model": "11",
            "name": "iPhone 11"
          },
          "dist": "abc123",
          "environment": "production",
          "feedback_id": "1ffe0775ac0f4417aed9de36d9f6f8dc",
          "locale": {
            "lang": "en",
            "timezone": "UTC+1"
          },
          "message": "I really like this feedback feature!",
          "os": {
            "name": "iOS",
            "version": "16.2"
          },
          "platform": "javascript",
          "project_id": "11276",
          "release": "version@1.3",
          "replay_id": "ec3b4dc8b79f417596f7a1aa4fcca5d2",
          "sdk": {
            "name": "sentry.javascript.react",
            "version": "6.18.1"
          },
          "status": "unresolved",
          "tags": {
            "hello": "is",
            "it": ["me", "you're", "looking", "for"]
          },
          "timestamp": "2023-08-31T14:10:34.954048",
          "url": "https://docs.sentry.io/platforms/javascript/",
          "user": {
            "display_name": "John Doe",
            "email": "john.doe@example.com",
            "id": "30246326",
            "ip": "213.164.1.114",
            "username": "John Doe"
          }
        }
      ]
    }
    ```

## Feedback [/projects/<organization_slug>/<project_slug>/feedback/<feedback_id>/]

- Model

  - Body

    ```json
    {
      "data": {
        "browser": {
          "name": "Chome",
          "version": "103.0.38"
        },
        "contact_email": "colton.allen@sentry.io",
        "device": {
          "brand": "Apple",
          "family": "iPhone",
          "model": "11",
          "name": "iPhone 11"
        },
        "dist": "abc123",
        "environment": "production",
        "feedback_id": "1ffe0775ac0f4417aed9de36d9f6f8dc",
        "locale": {
          "lang": "en",
          "timezone": "UTC+1"
        },
        "message": "I really like this feedback feature!",
        "os": {
          "name": "iOS",
          "version": "16.2"
        },
        "platform": "javascript",
        "project_id": "11276",
        "release": "version@1.3",
        "replay_id": "ec3b4dc8b79f417596f7a1aa4fcca5d2",
        "sdk": {
          "name": "sentry.javascript.react",
          "version": "6.18.1"
        },
        "status": "unresolved",
        "tags": {
          "hello": "is",
          "it": ["me", "you're", "looking", "for"]
        },
        "timestamp": "2023-08-31T14:10:34.954048",
        "url": "https://docs.sentry.io/platforms/javascript/",
        "user": {
          "display_name": "John Doe",
          "email": "john.doe@example.com",
          "id": "30246326",
          "ip": "213.164.1.114",
          "username": "John Doe"
        }
      }
    }
    ```

### Fetch Feedback [GET]

Retrieve a single feedback item.

- Response 200

  [UserFeedback]

### Update Feedback [PATCH]

Partially update a feedback item.

- Request

  ```json
  {
    "data": {
      "status": "resolved"
    }
  }
  ```

- Response 202

  [UserFeedback]

### Delete Feedback [DELETE]

Delete a feedback item.

- Response 204

## Feedback Ingest [/feedback/]

### Create Feedback [POST]

Create a new feedback item. This method is a subset of the Event protocol with
an additional "feedback" object added. Within the feedback object are feedback
related event metadata.

Every field not marked as "optional" is considered a required field and must be
present in the request body.

See https://develop.sentry.dev/sdk/event-payloads/types/ for more information

| Column                     | Type             | Description                           |
| -------------------------- | ---------------- | ------------------------------------- |
| dist                       | optional[string] | -                                     |
| environment                | optional[string] | -                                     |
| event_id                   | optional[string] | Omitted IDs are internally generated. |
| feedback                   | object           | -                                     |
| feedback.contact_email     | optional[string] | -                                     |
| feedback.message           | string           | -                                     |
| feedback.name              | optional[string] | -                                     |
| feedback.replay_id         | optional[string] | -                                     |
| feedback.url               | string           | -                                     |
| platform                   | string           | -                                     |
| release                    | optional[string] | -                                     |
| request                    | optional[object] | -                                     |
| request.headers            | optional[object] | -                                     |
| request.headers.User-Agent | optional[string] | -                                     |
| sdk                        | object           | -                                     |
| sdk.name                   | string           | -                                     |
| sdk.version                | string           | -                                     |
| tags                       | optional[object] | -                                     |
| timestamp                  | number           | UTC timestamp.                        |
| user                       | optional[object] | -                                     |
| user.email                 | optional[string] | -                                     |
| user.id                    | optional[string] | -                                     |
| user.ip_address            | optional[string] | -                                     |
| user.name                  | optional[string] | -                                     |
| user.username              | optional[string] | -                                     |

- Request

  ```json
  {
    "dist": "abc123",
    "environment": "production",
    "event_id": "1ffe0775ac0f4417aed9de36d9f6f8dc",
    "feedback": {
      "contact_email": "colton.allen@sentry.io",
      "message": "I really like this feedback feature!",
      "replay_id": "ec3b4dc8b79f417596f7a1aa4fcca5d2",
      "url": "https://docs.sentry.io/platforms/javascript/"
    },
    "platform": "javascript",
    "release": "version@1.3",
    "request": {
      "headers": {
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
      }
    },
    "sdk": {
      "name": "sentry.javascript.react",
      "version": "6.18.1"
    },
    "tags": {
      "key": "value"
    },
    "timestamp": "2023-08-31T14:10:34.954048",
    "user": {
      "email": "username@example.com",
      "id": "123",
      "ip_address": "127.0.0.1",
      "name": "user",
      "username": "user2270129"
    }
  }
  ```
