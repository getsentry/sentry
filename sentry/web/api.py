"""
sentry.web.views
~~~~~~~~~~~~~~~~

:copyright: (c) 2010 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
import logging

from django.http import HttpResponse, HttpResponseBadRequest, \
  HttpResponseForbidden
from django.views.decorators.csrf import csrf_exempt
from django.views.decorators.http import require_http_methods

from sentry.conf import settings
from sentry.coreapi import project_from_auth_vars, project_from_api_key_and_id, \
  project_from_id, decode_and_decompress_data, safely_load_json_string, \
  validate_data, insert_data_to_database, APIError, APIUnauthorized, \
  extract_auth_vars, InvalidTimestamp
from sentry.models import Group, GroupBookmark, Project, View, ProjectDomain
from sentry.utils import json
from sentry.utils.http import is_same_domain, apply_access_control_headers
from sentry.web.decorators import has_access
from sentry.web.frontend.groups import _get_group_list
from sentry.web.helpers import render_to_response, render_to_string

logger = logging.getLogger('sentry.errors.coreapi')


@csrf_exempt
@require_http_methods(['POST', 'OPTIONS'])
def store(request):
    """
    The primary endpoint for storing new events.

    This will validate the client's authentication and data, and if
    successfull pass on the payload to the internal database handler.

    Authentication works in three flavors:

    1. Explicit signed requests

       These are implemented using the documented signed request protocol, and
       require an authentication header which is signed using with the project
       member's secret key.

    2. Explicit trusted requests

       Generally used for communications with client-side platforms (such as
       JavaScript in the browser), they require the GET variables public_key
       and project_id, as well as an HTTP_REFERER to be set from a trusted
       domain.

    3. Implicit trusted requests

       Used by the Sentry core, they are only available from same-domain requests
       and do not require any authentication information. They only require that
       the user be authenticated, and a project_id be sent in the GET variables.

    """
    try:
        if request.method == 'POST':
            auth_vars = extract_auth_vars(request)
            data = request.raw_post_data

            if auth_vars:
                server_version = auth_vars.get('sentry_version', '1.0')
                client = auth_vars.get('sentry_client')
            else:
                server_version = request.GET.get('version', '1.0')
                client = request.META.get('HTTP_USER_AGENT', request.GET.get('client'))

            if server_version not in ('1.0', '2.0'):
                raise APIError('Client/server version mismatch: Unsupported version: %r' % server_version)

            if server_version != '1.0' and not client:
                raise APIError('Client request error: Missing client version identifier.')

            referrer = request.META.get('HTTP_REFERER')

            if auth_vars:
                project = project_from_auth_vars(auth_vars, data)
            elif request.GET.get('api_key') and request.GET.get('project_id'):
                # public requests only need referrer validation for CSRF
                project = project_from_api_key_and_id(request.GET['api_key'], request.GET['project_id'])
                if not ProjectDomain.test(project, referrer):
                    raise APIUnauthorized()
            elif request.GET.get('project_id') and request.user.is_authenticated() and \
                 is_same_domain(request.build_absolute_uri(), referrer):
                # authenticated users are simply trusted to provide the right id
                project = project_from_id(request)
            else:
                raise APIUnauthorized()

            if not data.startswith('{'):
                data = decode_and_decompress_data(data)
            data = safely_load_json_string(data)

            try:
                validate_data(project, data)
            except InvalidTimestamp:
                # Log the error, remove the timestamp, and revalidate
                logger.error('Client %r passed an invalid value for timestamp %r' % (
                    data['timestamp'],
                    client or '<unknown client>',
                ))
                del data['timestamp']
                validate_data(project, data)

            insert_data_to_database(data)
    except APIError, error:
        response = HttpResponse(error.msg, status=error.http_status)
    else:
        response = HttpResponse('')
    return apply_access_control_headers(response)


@csrf_exempt
@has_access
def notification(request, project):
    return render_to_response('sentry/partial/_notification.html', request.GET)


@csrf_exempt
@has_access
def poll(request, project):
    from sentry.templatetags.sentry_helpers import as_bookmarks, handle_before_events

    offset = 0
    limit = settings.MESSAGES_PER_PAGE

    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    filters, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list = list(event_list[offset:limit])
    handle_before_events(request, event_list)

    data = [
        {
            'id': m.pk,
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
                'is_bookmarked': b,
            }).strip(),
            'title': m.message_top(),
            'message': m.error(),
            'level': m.get_level_display(),
            'logger': m.logger,
            'count': m.times_seen,
            'score': getattr(m, 'sort_value', None),
        } for m, b in as_bookmarks(event_list, request.user)]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def resolve(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()
    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    Group.objects.filter(pk=group.pk).update(status=1)
    group.status = 1

    data = [
        (m.pk, {
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
            }).strip(),
            'count': m.times_seen,
        }) for m in [group]]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def bookmark(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()

    if not request.user.is_authenticated():
        return HttpResponseForbidden()

    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    gb, created = GroupBookmark.objects.get_or_create(
        project=group.project,
        user=request.user,
        group=group,
    )
    if not created:
        gb.delete()

    response = HttpResponse(json.dumps({'bookmarked': created}))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def clear(request, project):
    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    _, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list.update(status=1)

    data = []
    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def chart(request, project):
    gid = request.REQUEST.get('gid')
    days = int(request.REQUEST.get('days', '90'))

    if gid:
        try:
            group = Group.objects.get(pk=gid)
        except Group.DoesNotExist:
            return HttpResponseForbidden()

        data = Group.objects.get_chart_data(group, max_days=days)
    else:
        data = Project.objects.get_chart_data(project, max_days=days)

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response
