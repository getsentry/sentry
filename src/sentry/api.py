"""
sentry.api
~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

import base64
from django.contrib.auth.models import User, AnonymousUser
from tastypie import fields
from tastypie.authentication import Authentication
from tastypie.authorization import Authorization
from tastypie.api import Api
from tastypie.resources import ModelResource
from sentry.models import Project, Team, TeamMember, Group, Event, ProjectKey
from sentry.web.helpers import get_project_list


class SentryAuthentication(Authentication):
    def is_authenticated(self, request, **kwargs):
        """
        Simple basic auth where username is the public key, and password
        is the secret key.
        """
        auth = request.META.get('HTTP_AUTHORIZATION')
        if not auth:
            return False

        if not auth.startswith('Basic '):
            return

        try:
            decoded = base64.b64decode(auth.split(' ', 1)[-1])
            public, secret = decoded.split(':', 1)
        except Exception:
            return False

        try:
            apikey = ProjectKey.objects.get(public_key=public, secret_key=secret)
        except ProjectKey.DoesNotExist:
            return False

        request.projects = [apikey.project] or get_project_list(apikey.user)
        if apikey.user:
            request.teams = Team.objects.get_for_user(apikey.user)
        else:
            request.teams = []
        request.user = apikey.user or AnonymousUser()
        request.apikey = apikey

        return True

    def get_identifier(self, request):
        # XXX: no idea what this is for
        return request.apikey.id


class TeamAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(pk__in=[t.pk for t in request.teams])
        return object_list.none()


class TeamBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(team__project__in=request.projects)
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(pk__in=[p.id for p in request.projects])
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(project__team__in=request.teams)
        return object_list.none()


class UserResource(ModelResource):
    username = fields.CharField('username', readonly=True)
    # name = fields.CharField('first_name', readonly=True)

    class Meta:
        queryset = User.objects.all()
        detail_uri_name = 'username'
        include_resource_uri = False
        resource_name = 'user'
        detail_allowed_methods = []
        list_allowed_methods = []
        fields = ['username']
        authentication = SentryAuthentication()


class TeamResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    members = fields.ToManyField('sentry.api.TeamMemberResource', 'member_set', full=True)

    class Meta:
        queryset = Team.objects.all()
        detail_uri_name = 'slug'
        resource_name = 'team'
        detail_allowed_methods = ['get']
        list_allowed_methods = ['get']
        fields = ['id', 'name']
        authentication = SentryAuthentication()
        authorization = TeamAuthorization()


class TeamMemberResource(ModelResource):
    user = fields.ToOneField(UserResource, 'user', full=True)
    # access = fields.CharField('type')

    class Meta:
        queryset = TeamMember.objects.all()
        include_resource_uri = False
        detail_allowed_methods = ['get']
        list_allowed_methods = []
        fields = ['user']
        authentication = SentryAuthentication()
        authorization = TeamBasedAuthorization()

    # def dehydrate_access(self, bundle):
    #     return bundle.obj.get_type_display()


class ProjectResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    team = fields.ToOneField(TeamResource, 'team', full=True)

    class Meta:
        queryset = Project.objects.filter(status=0)
        detail_uri_name = 'slug'
        resource_name = 'project'
        fields = ['id', 'name']
        allowed_methods = ['get']
        authentication = SentryAuthentication()
        authorization = ProjectAuthorization()


class GroupResource(ModelResource):
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    team = fields.ToOneField(TeamResource, 'team', readonly=True)
    message = fields.CharField('message', readonly=True)
    times_seen = fields.IntegerField('times_seen', readonly=True)
    last_seen = fields.DateTimeField('last_seen', readonly=True)
    first_seen = fields.DateTimeField('first_seen', readonly=True)
    latest_event = fields.ToOneField('sentry.api.EventResource', 'latest_event', readonly=True, full=True)
    # events = fields.ToManyField('sentry.api.EventResource', 'event_set')

    class Meta:
        queryset = Group.objects.all()
        resource_name = 'group'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    def full_dehydrate(self, bundle):
        # HACK:
        bundle.obj.team = bundle.obj.project.team
        bundle.obj.latest_event = bundle.obj.get_latest_event()
        return super(GroupResource, self).full_dehydrate(bundle)


class EventResource(ModelResource):
    group = fields.ToOneField(GroupResource, 'group', readonly=True)
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    message = fields.CharField('message', readonly=True)
    # data = fields.DictField(readonly=True)
    datetime = fields.DateTimeField('datetime', readonly=True)

    class Meta:
        queryset = Event.objects.all()
        resource_name = 'event'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    # def dehydrate_data(self, bundle):
    #     return bundle.obj.data


v1_api = Api(api_name='v1')
v1_api.register(TeamResource())
v1_api.register(TeamMemberResource())
v1_api.register(ProjectResource())
v1_api.register(GroupResource())
v1_api.register(EventResource())
