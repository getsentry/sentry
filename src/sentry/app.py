"""
sentry.app
~~~~~~~~~~

:copyright: (c) 2010-2014 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
from __future__ import absolute_import

from threading import local

from django.conf import settings
from raven.contrib.django.models import client

from sentry.utils import redis
from sentry.utils.imports import import_string
from sentry.utils.locking.backends.redis import RedisLockBackend
from sentry.utils.locking.manager import LockManager


class State(local):
    request = None
    data = {}

env = State()


def get_instance(path, options):
    cls = import_string(path)
    return cls(**options)


# TODO(dcramer): this is getting heavy, we should find a better way to structure
# this
buffer = get_instance(settings.SENTRY_BUFFER, settings.SENTRY_BUFFER_OPTIONS)
digests = get_instance(settings.SENTRY_DIGESTS, settings.SENTRY_DIGESTS_OPTIONS)
quotas = get_instance(settings.SENTRY_QUOTAS, settings.SENTRY_QUOTA_OPTIONS)
nodestore = get_instance(
    settings.SENTRY_NODESTORE, settings.SENTRY_NODESTORE_OPTIONS)
ratelimiter = get_instance(
    settings.SENTRY_RATELIMITER, settings.SENTRY_RATELIMITER_OPTIONS)
search = get_instance(settings.SENTRY_SEARCH, settings.SENTRY_SEARCH_OPTIONS)
tsdb = get_instance(settings.SENTRY_TSDB, settings.SENTRY_TSDB_OPTIONS)
raven = client
locks = LockManager(RedisLockBackend(redis.clusters.get('default')))
