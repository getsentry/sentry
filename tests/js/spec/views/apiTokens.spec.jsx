import React from 'react';

import {mountWithTheme} from 'sentry-test/enzyme';

import {Client} from 'app/api';
import {ApiTokens} from 'app/views/settings/account/apiTokens';

const organization = TestStubs.Organization();

describe('ApiTokens', function() {
  const routerContext = TestStubs.routerContext();

  beforeEach(function() {
    Client.clearMockResponses();
  });

  it('renders empty result', function() {
    Client.addMockResponse({
      url: '/api-tokens/',
    });

    const wrapper = mountWithTheme(
      <ApiTokens organization={organization} />,
      routerContext
    );

    // Should be loading
    expect(wrapper).toSnapshot();
  });

  it('renders with result', function() {
    Client.addMockResponse({
      url: '/api-tokens/',
      body: [TestStubs.ApiToken()],
    });

    const wrapper = mountWithTheme(
      <ApiTokens organization={organization} />,
      routerContext
    );

    // Should be loading
    expect(wrapper).toSnapshot();
  });

  it('can delete token', function() {
    Client.addMockResponse({
      url: '/api-tokens/',
      body: [TestStubs.ApiToken()],
    });

    const mock = Client.addMockResponse({
      url: '/api-tokens/',
      method: 'DELETE',
    });

    expect(mock).not.toHaveBeenCalled();

    const wrapper = mountWithTheme(
      <ApiTokens organization={organization} />,
      routerContext
    );

    wrapper.find('button[aria-label="Remove"]').simulate('click');

    // Should be loading
    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith(
      '/api-tokens/',
      expect.objectContaining({
        method: 'DELETE',
      })
    );
  });
});
