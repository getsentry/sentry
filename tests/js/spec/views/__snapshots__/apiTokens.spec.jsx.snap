// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiTokens renders empty result 1`] = `
<SideEffect(DocumentTitle)
  title="API Tokens - Sentry"
>
  <div>
    <StyledSettingsPageHeading
      action={
        <ForwardRef
          data-test-id="create-token"
          priority="primary"
          size="small"
          to="/settings/account/api/auth-tokens/new-token/"
        >
          Create New Token
        </ForwardRef>
      }
      noTitleStyles={false}
      title="Auth Tokens"
    />
    <AlertLink
      openInNewTab={false}
      priority="warning"
      size="normal"
      to="/settings/org-slug/developer-settings/new-internal"
      withoutMarginBottom={false}
    >
      Auth Tokens are tied to the logged in user, meaning they'll stop working if the user leaves the organization! We suggest using internal integrations to create/manage tokens tied to the organization instead.
    </AlertLink>
    <TextBlock>
      Authentication tokens allow you to perform actions against the Sentry API on behalf of your account. They're the easiest way to get started using the API.
    </TextBlock>
    <TextBlock>
      <span
        key="5"
      >
        <span
          key="0"
        >
          For more information on how to use the web API, see our 
        </span>
        <a
          href="https://docs.sentry.io/hosted/api/"
          key="2"
        >
          <span
            key="1"
          >
            documentation
          </span>
        </a>
        <span
          key="3"
        >
          .
        </span>
      </span>
    </TextBlock>
    <TextBlock>
      <small>
        psst. Looking for the 
        <strong>
          DSN
        </strong>
         for an SDK? You'll find that under
         
        <strong>
          [Project] » Settings » Client Keys
        </strong>
        .
      </small>
    </TextBlock>
    <Panel>
      <PanelHeader>
        Auth Token
      </PanelHeader>
      <PanelBody
        flexible={false}
        withPadding={false}
      >
        <EmptyMessage>
          You haven't created any authentication tokens yet.
        </EmptyMessage>
      </PanelBody>
    </Panel>
  </div>
</SideEffect(DocumentTitle)>
`;

exports[`ApiTokens renders with result 1`] = `
<SideEffect(DocumentTitle)
  title="API Tokens - Sentry"
>
  <div>
    <StyledSettingsPageHeading
      action={
        <ForwardRef
          data-test-id="create-token"
          priority="primary"
          size="small"
          to="/settings/account/api/auth-tokens/new-token/"
        >
          Create New Token
        </ForwardRef>
      }
      noTitleStyles={false}
      title="Auth Tokens"
    />
    <AlertLink
      openInNewTab={false}
      priority="warning"
      size="normal"
      to="/settings/org-slug/developer-settings/new-internal"
      withoutMarginBottom={false}
    >
      Auth Tokens are tied to the logged in user, meaning they'll stop working if the user leaves the organization! We suggest using internal integrations to create/manage tokens tied to the organization instead.
    </AlertLink>
    <TextBlock>
      Authentication tokens allow you to perform actions against the Sentry API on behalf of your account. They're the easiest way to get started using the API.
    </TextBlock>
    <TextBlock>
      <span
        key="5"
      >
        <span
          key="0"
        >
          For more information on how to use the web API, see our 
        </span>
        <a
          href="https://docs.sentry.io/hosted/api/"
          key="2"
        >
          <span
            key="1"
          >
            documentation
          </span>
        </a>
        <span
          key="3"
        >
          .
        </span>
      </span>
    </TextBlock>
    <TextBlock>
      <small>
        psst. Looking for the 
        <strong>
          DSN
        </strong>
         for an SDK? You'll find that under
         
        <strong>
          [Project] » Settings » Client Keys
        </strong>
        .
      </small>
    </TextBlock>
    <Panel>
      <PanelHeader>
        Auth Token
      </PanelHeader>
      <PanelBody
        flexible={false}
        withPadding={false}
      >
        <ApiTokenRow
          key="apitoken123"
          onRemove={[Function]}
          token={
            Object {
              "dateCreated": 2018-01-12T02:01:41.000Z,
              "scopes": Array [
                "scope1",
                "scope2",
              ],
              "token": "apitoken123",
            }
          }
        />
      </PanelBody>
    </Panel>
  </div>
</SideEffect(DocumentTitle)>
`;
